api/include/opentelemetry/http/http_client.h

```diff
@@ -0,0 +1,161 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include "opentelemetry/nostd/string_view.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace http
+{
+
+enum class HttpResult : uint8_t
+{
+  // Response has been received successfully from target server.
+  HttpResult_OK = 0,
+
+  // Request has been aborted by the caller. The server might or
+  // might not have already received or processed the request.
+  HttpResult_Aborted = 1,
+
+  // Local conditions have prevented the request from being sent
+  // (invalid request parameters, out of memory, internal error etc.)
+  HttpResult_LocalFailure = 2,
+
+  // Network conditions somewhere between the local machine and
+  // the target server have caused the request to fail
+  // (connection failed, connection dropped abruptly etc.).
+  HttpResult_NetworkFailure = 3
+
+};
+
+// The HttpHeaders class contains a set of HTTP headers.
+class HttpHeaders
+{
+public:
+  // Inserts a name/value pair, and removes elements with the same name.
+  virtual void set(nostd::string_view const &name, nostd::string_view const &value) = 0;
+
+  // Inserts a name/value pair
+  virtual void add(nostd::string_view const &name, nostd::string_view const &value) = 0;
+
+  // Gets a string value given a name
+  // Returns:
+  //      If there are multiple headers with same name, it returns any one of the value
+  //      (implementation defined) Empty string if there is no header with speicfied name..
+  virtual nostd::string_view const &get(nostd::string_view const &name) const = 0;
+
+  // Tests whether the headers contain the specified name.
+  virtual bool contains(nostd::string_view const &name) const = 0;
+};
+
+// The HttpRequest class represents a Request object.
+// Individual HTTP client implementations can implement the request object in
+// the most efficient way. Either fill the request first, and then issue
+// the underlying request, or create the real request immediately, and then forward
+// the methods and set the individual parameters directly one by one.
+class HttpRequest
+{
+public:
+  // Gets the request ID.
+  virtual const nostd::string_view &GetId() const = 0;
+
+  // Sets the request Method
+  virtual void SetMethod(nostd::string_view const &method) = 0;
+
+  // Gets the request URI
+  virtual void SetUrl(nostd::string_view const &url) = 0;
+
+  // Gets the request Headers
+  virtual HttpHeaders &GetHeaders() const = 0;
+
+  // Sets the request body
+  virtual void SetBody(const uint8_t *const body, const size_t len) = 0;
+
+  // Gets the request body
```

Comment:
How does the caller know `GetBody` is succeeded, check non zero `len` after return? If so, perhaps mention this in the comment.

Reply:
`GetBody` gette shouldn't fail. It will return zero in `len` if body is not set, else return the actual size. Both are still the success scenarios. Anyway, I am converting it to vector<uint8_t> and would get more clear :)